test/extensions/common/aws/metadata_fetcher_test.cc

```diff
@@ -0,0 +1,301 @@
+      .WillOnce(Invoke([](Http::RequestMessagePtr& request, Http::AsyncClient::Callbacks&,
+                          const Http::AsyncClient::RequestOptions&) -> Http::AsyncClient::Request* {
+        EXPECT_EQ("169.254.170.2", request->headers().getHostValue());
+        EXPECT_EQ("/v2/credentials/c68caeb5-ef71-4914-8170-111111111111",
+                  request->headers().getPathValue());
+        EXPECT_EQ(Http::Headers::get().MethodValues.Get, request->headers().getMethodValue());
+        EXPECT_EQ(Http::Headers::get().SchemeValues.Http, request->headers().getSchemeValue());
+        EXPECT_EQ("Token", request->headers()
+                               .get(Http::LowerCaseString("X-aws-ec2-metadata-token"))[0]
+                               ->value()
+                               .getStringView());
+        EXPECT_TRUE(request->headers().get(Http::LowerCaseString(":pseudo-header")).empty());
+        return nullptr;
+      }));
+
+  // Act
+  fetcher_->fetch(message, parent_span_, receiver);
+}
+
+TEST_F(MetadataFetcherTest, TestGet400) {
+  // Setup
+  setupFetcher();
+  Http::RequestMessageImpl message;
+
+  MockUpstream mock_result(mock_factory_ctx_.cluster_manager_, "400", "not_empty");
+  MockMetadataReceiver receiver;
+  EXPECT_CALL(receiver, onMetadataSuccess(testing::_)).Times(0);
+  EXPECT_CALL(receiver, onMetadataError(MetadataFetcher::MetadataReceiver::Failure::Network));
+
+  // Act
+  fetcher_->fetch(message, parent_span_, receiver);
+}
+
+TEST_F(MetadataFetcherTest, TestGet400NoBody) {
+  // Setup
+  setupFetcher();
+  Http::RequestMessageImpl message;
+
+  MockUpstream mock_result(mock_factory_ctx_.cluster_manager_, "400", "");
+  MockMetadataReceiver receiver;
+  EXPECT_CALL(receiver, onMetadataSuccess(testing::_)).Times(0);
+  EXPECT_CALL(receiver, onMetadataError(MetadataFetcher::MetadataReceiver::Failure::Network));
+
+  // Act
+  fetcher_->fetch(message, parent_span_, receiver);
+}
+
+TEST_F(MetadataFetcherTest, TestGetNoBody) {
+  // Setup
+  setupFetcher();
+  Http::RequestMessageImpl message;
+
+  MockUpstream mock_result(mock_factory_ctx_.cluster_manager_, "200", "");
+  MockMetadataReceiver receiver;
+  EXPECT_CALL(receiver, onMetadataSuccess(testing::_)).Times(0);
+  EXPECT_CALL(receiver,
+              onMetadataError(MetadataFetcher::MetadataReceiver::Failure::InvalidMetadata));
+
+  // Act
+  fetcher_->fetch(message, parent_span_, receiver);
+}
+
+TEST_F(MetadataFetcherTest, TestHttpFailure) {
+  // Setup
+  setupFetcher();
+  Http::RequestMessageImpl message;
+
+  MockUpstream mock_result(mock_factory_ctx_.cluster_manager_,
+                           Http::AsyncClient::FailureReason::Reset);
+  MockMetadataReceiver receiver;
+  EXPECT_CALL(receiver, onMetadataSuccess(testing::_)).Times(0);
+  EXPECT_CALL(receiver, onMetadataError(MetadataFetcher::MetadataReceiver::Failure::Network));
+
+  // Act
+  fetcher_->fetch(message, parent_span_, receiver);
+}
+
+TEST_F(MetadataFetcherTest, TestWithValidCluster) {
```

Comment:
Could you elaborate in comments or the test name what's expected to happen here? I find it surprising that the expectation is "onMetadataError is not called, but also, implicitly, onMetadataSuccess is not called."